### PR TITLE
Notional Bridge

### PR DESCRIPTION
**Pull request by haowang0402**
# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] There are no unexpected formatting changes, or superfluous debug logs.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewers next convenience.
- [ ] NatSpec documentation of all the non-test functions is present and is complete.
- [ ] Continuous integration (CI) passes.

### CODE REVIEW

**Comment by benesjan** [comment]
@haowang0402 if the bridge is ready for review please request in the Reviewers tab on the right.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:1]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
// Copyright 2022 Aztec.
```

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:16]: fcashFactory can be set as constant since it's a contract deployed on mainnet.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:62]: Missing description of return variable.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:116]: Missing description of return variable.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:144]: Missing description of return variable.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:181]: Missing description of return variable.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:90]: This check is not necessary because we can trust that RollupProcessor sends enough ETH if it's an input.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:48]: Setting currencyIds in constructor is not a good idea because it can happen that Notional adds support for another cToken and then the bridge could not support that. Doesn't notional have a contract similar to Compound's Comptroller which tracks all the supported markets? Check our [Compound bridge](https://github.com/AztecProtocol/aztec-connect-bridges/blob/a233b8ebd5e8648c100c86de430532bffc3576cf/src/bridges/compound/CompoundBridge.sol#L53) for how we handle it there.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:89]: Checking for ETH should be done only by asset type because it could happen that the asset would be of type `NOT_USED` and then this condition would pass as well (structs are initialized with 0 values when not specified otherwise)..

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:105]: Since the bridge doesn't hold any tokens it would be safe to pre-approve the tokens. This has the advantage of significantly decreasing gas costs. This would be ideally done in the same function as in which you will load the currencyIds. Check [compound](https://github.com/AztecProtocol/aztec-connect-bridges/blob/a233b8ebd5e8648c100c86de430532bffc3576cf/src/bridges/compound/CompoundBridge.sol#L53) for an inspiration.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:146]: Please remove this check. Once the approval will be only in the separate "approvals" function this method would revert anyway in case there would be inocorrect token. With this in mind this check only wastes gas.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:159]: In what situations would this condition happen? If fcash was not deployed then you could not receive currencyId from it, no?

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:49]: Please generalise this comment to not list the underlying exactly - Notional could add market.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:106]: Missing description of _underlyingToken.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:192]: The bridge is not expected to hold any tokens between interaction so it would be safe to just return full balance to the users. If someone send some tokens to the bridge randomly, those users will be lucky. For this reason please remove the prevBalance variable.

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalE2E.t.sol:46]: Please replace all the `vm.assume(...)` calls you are using in the tests with bound(...) form forge-std. This change is good to do because `vm.assume(...)` is not ideal when the amount of excluded values are too high --> the fuzzer then executes too few tests. Another tip is that you can make `_depositAmount` be a smaller type than is `uint256` (e.g. `uint64`). This way you can tell the fuzzer to use smaller values as well.

**Review by benesjan — CHANGES_REQUESTED** [review]

File: src/bridges/notional/NotionalBridge.sol
Comment (haowang0402) [reply src/bridges/notional/NotionalBridge.sol:27]: In the function `convert`, we will approve `rollupProcessor` with `type(uint).max` to transfer the token if it has not been approved. Do you think we still need an extra callable function in that case?

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:164]: I think it would be cleaner and less bug prone if you simply passed to these functions AztecAsset instead of the ERC20 address. Then you could check whether the asset is ETH by checking if assetType is equal to AztecAssetType.ETH.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:132]: I think it would be cleaner and less bug prone if you simply passed to these functions AztecAsset instead of the ERC20 address. Then you could check whether the asset is ETH by checking if assetType is equal to AztecAssetType.ETH.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:117]: Outdated NatSpec. You are not setting any approvals anymore.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:88]:
```suggestion
     * @notice Helper function to compute the fcash amount
```

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalE2E.t.sol:18]: Unused storage variable.

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalE2E.t.sol:16]: Comment referencing other bridge

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalE2E.t.sol:47]: Why is it necessary to deploy the wrapper here? Isn't some already deployed on mainnet?

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalE2E.t.sol:84]: Please also add E2E tests of ERC20 tokens. This is necessary so that we properly test approvals and gas limits.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:200]: Broken name.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:351]: Please generalize these tests so that they fetch tokens form Notional view contract and iteratively runs these tests in a for loop.

**Review by benesjan — CHANGES_REQUESTED** [review]
Please write tests of the norional-bridge-data class. Also please rebase the branch onto master and then force push the changes. There is a tutorial linked in the README. The PR should not touch any other files than the Notional related ones.

File: src/bridges/notional/NotionalBridge.sol
Comment (haowang0402) [reply src/bridges/notional/NotionalBridge.sol:35]: For the previous commit, you mentioned that we don't want the mapping `preapproved`? Also with `perApprovalAll()`, we still need to call approve for `fcash` in the `_enter` function.

File: src/test/bridges/notional/NotionalE2E.t.sol
Comment (haowang0402) [reply src/test/bridges/notional/NotionalE2E.t.sol:47]: There are two fcash contracts deployed on the mainnet. I can write e2e tests for those two instead.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [reply src/bridges/notional/NotionalBridge.sol:35]: We don't need to call approve on fcash in enter function if we simply set all the possible approvals in `perApprovalAll()`. Goal is that no approve has to be called within the execution of the `convert(...)` function. In the Compound bridge is an example. All the approvals set by calling 1 function.

Comment (haowang0402) [reply src/bridges/notional/NotionalBridge.sol:35]: I don't think it's possible since there is a fcash contract for each corresponding maturity. After the maturity, there will be new fcash contracts.

Comment (benesjan) [reply src/bridges/notional/NotionalBridge.sol:35]: Ok, makes sense. Didn't realize fcash tokens are completely new for each maturity. In that case it makes sense to have only pre-approvals of cTokens and underlying with spender being the rollup processor in this function. For that you can reuse parts of the Compound bridge code. For the rest it should stay in the convert function. Sorry for the confusion

**Comment by benesjan** [comment]
@haowang0402 before sending it for a review again please address all my comments and rebase onto master. Thank you

**Comment by benesjan** [comment]
Hello @haowang0402 , this PR contains changes unrelated to Notional. Please try rebasing onto master and force pushing to this branch. See [this article](https://blog.verslu.is/git/git-rebase/) for how to do it. If you have issues with that just copy the files to a new branch and submit a new PR. Please follow the PR checklist.
image

**Comment by haowang0402** [comment]
Hi, I just rebased to the master branch. Could you please take another look?

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:37]: there is emptyAsset in BridgeTestBase so this is redundant.

File: src/bridges/notional/NotionalBridge.sol
Comment (benesjan) [inline src/bridges/notional/NotionalBridge.sol:107]: Why is the amount uint88 and not the default uin256?

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:67]: Why are the lower bounds so different? What errors get thrown? Wouldn't it make sense to just have 1e16 everywhere?

**Review by benesjan — CHANGES_REQUESTED** [review]
Didn't have time today to finish full review. Will continue next week.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (haowang0402) [reply src/test/bridges/notional/NotionalUnit.t.sol:67]: Different tokens have different decimals, so we need to have different lower bounds. If it's below the lower bound, we would have a `liquidity failed` error.

File: src/bridges/notional/NotionalBridge.sol
Comment (haowang0402) [reply src/bridges/notional/NotionalBridge.sol:107]: Notional uses `uint88` for its internal tracking.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:27]: You don't need to cast address to address + it feels redundant that you have the addresses here and than you copy them to `tokens` array.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:84]: You don't need this mapping because all the cTokens have underlying method on them which gives you the address.

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:59]: Can't you just populate this dynamically? Some of notional contracts should return the id for address, no?

File: src/test/bridges/notional/NotionalUnit.t.sol
Comment (benesjan) [inline src/test/bridges/notional/NotionalUnit.t.sol:51]: You can just set this directly when initializing the array.

**Review by benesjan — CHANGES_REQUESTED** [review]
The tests are hard to read. Please add some spaces and comments there.

**Comment by benesjan** [comment]
Grant closed

**Pull request closed without merging**
